jerry-core/ecma/operations/ecma-typedarray-object.c

```diff
@@ -0,0 +1,941 @@
+ * Create a TypedArray object by given array_length
+ *
+ * See also: ES2015 22.2.1.2.1
+ *
+ * @return pointer to the new typedarray object
+ */
+static ecma_object_t *
+ecma_typedarray_create_object_with_length (ecma_length_t array_length, /**< length of the typedarray */
+                                           ecma_object_t *proto_p, /**< prototype object */
+                                           uint8_t element_size_shift, /**< the size shift of the element length */
+                                           lit_magic_string_id_t class_id) /**< class name of the typedarray */
+{
+  ecma_length_t byte_length = array_length << element_size_shift;
+  ecma_object_t *object_p = ecma_create_object (proto_p,
+                                                sizeof (ecma_extended_object_t),
+                                                ECMA_OBJECT_TYPE_CLASS);
+
+  ecma_extended_object_t *ext_object_p = (ecma_extended_object_t *) object_p;
+  ext_object_p->u.class_prop.class_id = class_id;
+  ext_object_p->u.class_prop.class_type = ECMA_CLASS_TYPE_TYPEDARRAY;
+  ext_object_p->u.class_prop.extra_info = element_size_shift;
+  ext_object_p->u.class_prop.u.value = ecma_make_object_value (ecma_arraybuffer_new_object (byte_length));
+  ecma_free_value (ext_object_p->u.class_prop.u.value);
+
+  return object_p;
+} /* !ecma_typedarray_create_object_with_length */
+
+/**
+ * Create a TypedArray object by given buffer, offset, and array_length
+ *
+ * See also: ES2015 22.2.1.5
+ *
+ * @return pointer to the new typedarray object
+ */
+static ecma_object_t *
+ecma_typedarray_create_object_with_buffer (ecma_object_t *arraybuffer_p,
+                                           ecma_length_t byte_offset,
+                                           ecma_length_t array_length, /**< length of the typedarray */
+                                           ecma_object_t *proto_p, /**< prototype object */
+                                           uint8_t element_size_shift, /**< the size shift of the element length */
+                                           lit_magic_string_id_t class_id) /**< class name of the typedarray */
+{
+  ecma_object_t *object_p;
+  if (byte_offset == 0 && array_length == ecma_arraybuffer_get_length (arraybuffer_p) >> element_size_shift)
+  {
+    object_p = ecma_create_object (proto_p,
+                                   sizeof (ecma_extended_object_t),
+                                   ECMA_OBJECT_TYPE_CLASS);
+
+    ecma_extended_object_t *ext_object_p = (ecma_extended_object_t *) object_p;
+    ext_object_p->u.class_prop.class_id = class_id;
+    ext_object_p->u.class_prop.class_type = ECMA_CLASS_TYPE_TYPEDARRAY;
+    ext_object_p->u.class_prop.extra_info = element_size_shift;
+    ext_object_p->u.class_prop.u.value = ecma_make_object_value (arraybuffer_p);
+  }
+  else
+  {
+    object_p = ecma_create_object (proto_p,
+                                   sizeof (ecma_extended_object_t) + sizeof (ecma_typedarray_info_t),
```

Comment:
I can use another struct which contains the `extended_object_t` and `typedarray_info_t`, but I don't think it is necessary in this case.

If there is a container struct, let's say `extended_typedarray_t` , we cannot say `sizeof (extended_typedarray_t) === sizeof(extended_object_t) + sizeof(typedarray_info_t)` because of the padding.

Now, it didn't allocate a container struct , but allocate a section of memory. We define the former `sizeof(ecma_extended_object_t)` length part is to store ecma_extended_object_t, and the remaining part (whose length is sizeof(ecma_typedarray_info_t)) is to store ecma_typedarray_info_t.

Please correct me if I missed anything.